### PR TITLE
Feat(tests): Add comprehensive unit tests for analysis and data fetching

### PR DESCRIPTION
**Pull request by mohavro**
## Description

Add comprehensive unit tests for analysis and data fetching. New PR due to problems with size of PR and head commit issues.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [X] Code refactoring
- [ ] Performance improvement
- [ ] Test addition/update
- [ ] Dependency update

## Related Issues

Closes #750 #748 #746 #744 #738

## Changes Made

- Added comprehensive unit test suite for FormulaicAnalyzer in tests/unit/test_formulaic_analysis.py.
- Added comprehensive unit test suite for RealDataFetcher in tests/unit/test_real_data_fetcher.py.
- Replaced unused key variable with _ in dict iteration at line 322
- Removed dead code blocks containing unused variable definitions and empty function stubs
- Refactored security tests in tests/integration/test_pr_agent_config_validation.py for improved clarity and maintainability.
- Applied minor code formatting across several files.

## Testing

Minimal checkout testing and use crucial tests only to avoid exhausting CI / workflow resources.

More extensive testing can be done when all PR tasks complete

- [X] Unit tests pass locally
- [X] Integration tests pass locally
- [ ] Manual testing completed
- [ ] New tests added for new functionality

### Test Commands

```bash
# Commands used to test
pytest
npm test
```

### Code Quality

- [X] My code follows the project's style guidelines
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Documentation

- [ ] I have updated the README.md (if needed)
- [ ] I have updated the CHANGELOG.md
- [ ] I have updated relevant documentation files
- [X] I have added/updated docstrings

### Dependencies

- [X] I have checked for dependency conflicts
- [ ] I have updated requirements.txt (if needed)
- [ ] I have updated package.json (if needed)

### Branch Management

- [X] My branch is up to date with main
- [X] I will delete this branch after merge
- [X] This branch has a descriptive name
- [X] This PR has been open for less than 2 weeks

## Additional Notes

**For Reviewers**: Please ensure the branch is deleted after merging this PR to maintain repository hygiene.

**Related Documentation**:
- [Contributing Guidelines](../CONTRIBUTING.md)
- [Branch Cleanup Guidelines](../BRANCH_CLEANUP_QUICK_REFERENCE.md)

### **Description**
- Added comprehensive unit tests for `FormulaicAnalyzer` and `RealDataFetcher`, enhancing test coverage and reliability.
- Improved configuration validation tests for the PR agent, ensuring better handling of sensitive information and YAML structure.
- Enhanced error handling in the `RealDataFetcher` to ensure cache directories are created as needed.

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_formulaic_analysis.py</strong><dd><code>Comprehensive Unit Tests for FormulaicAnalyzer</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_formulaic_analysis.py
<li>Added comprehensive unit tests for <code>FormulaicAnalyzer</code>.<br> <li> Enhanced tests for formula extraction and analysis.<br> <li> Improved test coverage for graph analysis methods.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/752/files#diff-e7c156c1adacce6d815a72c4f0d22dfb0fe1873dcced0784e3e7d3ca02170b04">+995/-480</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_real_data_fetcher.py</strong><dd><code>Comprehensive Unit Tests for RealDataFetcher</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_real_data_fetcher.py
<li>Added comprehensive unit tests for <code>RealDataFetcher</code>.<br> <li> Covered various configurations and data fetching methods.<br> <li> Included tests for edge cases and error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/752/files#diff-3181035937ccf7dfbd5be2364047453639e0969e14492c465612c07f2a17fa1e">+1230/-0</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_pr_agent_config_validation.py</strong><dd><code>Enhanced PR Agent Config Validation Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/integration/test_pr_agent_config_validation.py
<li>Improved configuration validation tests.<br> <li> Added checks for duplicate keys and consistent indentation.<br> <li> Enhanced security checks for sensitive information.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/752/files#diff-79f470e2eb177b55f768d16530ca72370defa508536fc0b09cab9d483aca0c2e">+229/-311</a></td>
</tr>
</table></td></tr><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>real_data_fetcher.py</strong><dd><code>Cache Directory Creation in RealDataFetcher</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/data/real_data_fetcher.py
<li>Added cache directory creation to ensure cache persistence.<br> <li> Improved error handling during cache operations.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/752/files#diff-eba7679b8e04549cd18d2006d101d6cf59e37e1e1eb1b63c84d5d4589fc617b1">+6/-3</a>&nbsp; &nbsp; &nbsp; </td>
</tr>
</table></td></tr></tr></tbody></table>

> 💡 **Penify usage**:
>Comment `/help` on the PR to get a list of all available Penify tools and their descriptions

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR adds comprehensive unit test coverage for analysis and data fetching components. The new test files (`test_formulaic_analysis.py` and `test_real_data_fetcher.py`) provide thorough coverage with proper mocking, fixtures, and edge case handling. Additionally, `test_repository.py` expands with complex real-world scenarios.

**Key Changes:**
- Added 1,230 lines of tests for `RealDataFetcher` covering initialization, caching, network failures, and all fetch methods
- Expanded `test_formulaic_analysis.py` to 1,192 lines with comprehensive formula extraction and analysis tests
- Enhanced `test_repository.py` with 500+ new lines testing portfolio workflows and edge cases
- Added input validation to `repository.py` for relationship strength values (0.0-1.0 range)
- Refactored security validation tests in `test_pr_agent_config_validation.py` with improved entropy-based heuristics
- Applied code formatting improvements across multiple files

**Issues Found:**
- `src/data/repository.py` has a duplicate docstring on the `get_relationship` method (lines 149-157)
- `src/data/repository.py` contains an incomplete placeholder comment (lines 106-107)

<h3>Confidence Score: 4/5</h3>

- Safe to merge after fixing the duplicate docstring in repository.py
- The PR adds excellent test coverage with proper mocking and edge case handling. The new tests are well-structured and comprehensive. However, there are two minor syntax issues in repository.py (duplicate docstring and incomplete comment) that should be cleaned up before merging. No logic errors or security concerns were found.
- src/data/repository.py needs cleanup for duplicate docstring and placeholder comment

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/data/repository.py | Added validation and docstrings, but has duplicate docstring and incomplete comment |
| tests/unit/test_formulaic_analysis.py | Comprehensive test suite expansion with proper fixtures and edge cases - excellent coverage |
| tests/unit/test_real_data_fetcher.py | New comprehensive test suite with mocking, covers all RealDataFetcher methods - thorough |
| tests/unit/test_repository.py | Expanded test suite with complex scenarios and edge cases - comprehensive coverage |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Dev as Developer
    participant Test as Test Suite
    participant Analyzer as FormulaicAnalyzer
    participant Fetcher as RealDataFetcher
    participant Repo as AssetGraphRepository
    participant Graph as AssetRelationshipGraph
    
    Note over Dev,Graph: PR #752: Comprehensive Unit Test Addition
    
    Dev->>Test: Add test_formulaic_analysis.py
    Test->>Analyzer: Test initialization
    Test->>Analyzer: Test analyze_graph()
    Analyzer->>Graph: Extract formulas from assets
    Graph-->>Analyzer: Return asset data
    Analyzer-->>Test: Return analysis results
    
    Dev->>Test: Add test_real_data_fetcher.py
    Test->>Fetcher: Test create_real_database()
    Fetcher->>Fetcher: Check cache_path
    alt Cache exists
        Fetcher->>Fetcher: Load from cache
    else Network enabled
        Fetcher->>Fetcher: Fetch from Yahoo Finance
    else Fallback
        Fetcher->>Fetcher: Use sample data
    end
    Fetcher->>Graph: Build graph with assets
    Graph-->>Fetcher: Return populated graph
    Fetcher-->>Test: Return graph
    
    Dev->>Test: Expand test_repository.py
    Test->>Repo: Test add_or_update_relationship()
    Repo->>Repo: Validate strength (0.0-1.0)
    alt Invalid strength
        Repo-->>Test: Raise ValueError
    else Valid
        Repo->>Graph: Store relationship
        Graph-->>Test: Success
    end
    
    Note over Dev,Graph: All tests include edge cases and error handling
```

___

## **CodeAnt-AI Description**
Revamp and expand unit tests for analysis, data fetching, and database behavior

### What Changed
- Replaced and expanded tests for FormulaicAnalyzer to validate outputs for equities, bonds, commodities, currencies, portfolio/risk/valuation formulas, formula categories, formula counts, LaTeX presence, and r_squared bounds; includes many edge-case scenarios (zero/negative prices, high P/E, very low prices, regulatory events, multi-asset graphs).
- Added a comprehensive test suite for RealDataFetcher that verifies initialization options, network-disabled fallback behavior, cache save/load roundtrips, serialization/deserialization of assets/events/graphs, fetching logic for equities/bonds/commodities/currencies (including error handling), and that cache save failures do not block returning a usable graph.
- Improved database tests to assert in-memory and URI-style SQLite persistence and behaviors (schema/data persistence across connections, memory-DB detection rules, connection row factory and threading safety, and that DATABASE_URL is required), with clearer expectations about persistence vs connection identity.

### Impact
`✅ Clearer regression detection for formula extraction`
`✅ Fewer cache and fetch-related regressions`
`✅ More reliable in-memory database behavior tests`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/ca226731-8bae-4be3-bc83-780ed57c75ab?refs=6fdb4e2d0d26844d59f68e34e2f274a398da3b4b&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: tests/unit/test_repository.py [hunk 1]
```diff
@@ -768,9 +792,511 @@ def test_relationship_with_max_strength(repository):
...
+        repository.session.commit()
+
+        repository.add_or_update_relationship(
+            "NEG1", "NEG2", "negative_corr", -0.8, bidirectional=False
+        )
+        repository.session.commit()
+
+        rel = repository.get_relationship("NEG1", "NEG2", "negative_corr")
+        assert rel is not None
+        assert rel.strength == -0.8
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_repository.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/752#discussion_r2748301863)

File: tests/unit/test_repository.py [hunk 2]
```diff
@@ -768,9 +792,511 @@ def test_relationship_with_max_strength(repository):
...
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_repository.py]:
```suggestion
    def test_negative_strength_relationship(repository):
        """Test relationship with negative strength (negative correlation)."""
        asset1 = Equity(
            id="NEG1",
            symbol="N1",
            name="Neg 1",
            asset_class=AssetClass.EQUITY,
            sector="Tech",
            price=100.0,
        )
        asset2 = Equity(
            id="NEG2",
            symbol="N2",
            name="Neg 2",
            asset_class=AssetClass.EQUITY,
            sector="Finance",
            price=200.0,
        )

        repository.upsert_asset(asset1)
        repository.upsert_asset(asset2)
        repository.session.commit()

        with pytest.raises(ValueError):
            repository.add_or_update_relationship(
                "NEG1", "NEG2", "negative_corr", -0.8, bidirectional=False
            )

```

File: src/analysis/formulaic_analysis.py [hunk 3]
```diff
@@ -459,3 +462,216 @@ def _calculate_avg_correlation_strength_from_empirical(
...
+                and hasattr(asset, "pe_ratio")
+                and asset.pe_ratio is not None
+            ):
+                examples.append(f"{asset.symbol}: PE = {asset.pe_ratio:.2f}")
+                if len(examples) >= 2:
+                    break
+        return (
+            "; ".join(examples)
+            if examples
+            "; ".join(examples)
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/analysis/formulaic_analysis.py]:
```suggestion
            return "; ".join(examples) if examples else "Example: PE = 150 / 10 = 15.0"
```

File: tests/unit/test_real_data_fetcher.py [hunk 4]
```diff
@@ -0,0 +1,1246 @@
...
+        assert "relationships" in serialized
+        assert "incoming_relationships" in serialized
+        # Bidirectional relationship creates entries in both
+        assert "E1" in serialized["relationships"]
+        assert "E2" in serialized["relationships"]
+
+
+class TestCacheOverwriteOperations:
+    """Test cache overwrite operations squentially."""
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_real_data_fetcher.py]:
```suggestion
class TestCacheOverwriteOperations:
    """Test cache overwrite operations sequentially."""

```

File: tests/unit/test_real_data_fetcher.py [hunk 5]
```diff
@@ -0,0 +1,1218 @@
...
+        }
+
+        event = _deserialize_event(data)
+
+        assert isinstance(event, RegulatoryEvent)
+        assert event.id == "EVENT1"
+        assert event.asset_id == "TEST"
+        assert event.event_type == "Earnings Report"
+        assert event.description == "Test event"
+        assert event.impact_score == 0.5
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_real_data_fetcher.py]:
```suggestion
    def test_deserialize_event(self):
        """Test deserializing a regulatory event."""
        data = {
            "id": "EVENT1",
            "asset_id": "TEST",
            "event_type": "Earnings Report",
            "date": "2024-01-01",
            "description": "Test event",
            "impact_score": 0.5,
        }

        event = _deserialize_event(data)

        assert isinstance(event, RegulatoryEvent)
        assert event.id == "EVENT1"
        assert event.asset_id == "TEST"
        assert event.event_type == RegulatoryActivity.EARNINGS_REPORT
        assert event.description == "Test event"
        assert event.impact_score == 0.5
```

**Review by mohavro — COMMENTED** [review]
approved. merge pr

**Review by mohavro — COMMENTED** [review]
approved

**Review by mohavro — COMMENTED** [review]
Resolve all unresolved comments.

**Review by mohavro — COMMENTED** [review]
approved

File: tests/integration/test_pr_agent_config_validation.py [hunk 6]
```diff
@@ -179,336 +307,143 @@ class TestPRAgentConfigSecurity:
...
+        """Assert numeric configuration limits fall within safe bounds."""
         limits = pr_agent_config["limits"]
 
-        # Check for reasonable numeric limits
-        assert limits["max_execution_time"] <= 3600, "Execution time too high"
-        assert limits["max_concurrent_prs"] <= 10, "Too many concurrent PRs"
-        assert limits["rate_limit_requests"] <= 1000, "Rate limit too high"
+        assert limits["max_execution_time"] <= 3600
+        assert limits["max_concurrent_prs"] <= 10
+        assert limits["rate_limit_requests"] <= 1000
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_pr_agent_config_validation.py]: @codeant-ai can you review the following and add type hints to all python code in this file?

In `@tests/integration/test_pr_agent_config_validation.py` around lines 382 - 427,
The tests and helper functions lack type hints and full docstrings; update
test_no_hardcoded_secrets and test_safe_configuration_values to declare the
pr_agent_config parameter type (e.g., pr_agent_config: dict), add explicit
return type annotations (-> None) for the test functions, and annotate the
internal helpers is_allowed_placeholder(v: object) -> bool and
scan_for_secrets(node: object, path: str = "root") -> None; also expand each
docstring (for the two test methods and the helper functions) to include a short
description plus Returns and Raises sections documenting return values and any
assertions/errors raised (e.g., Returns: bool/None; Raises: AssertionError for
detected secrets). Ensure you reference the existing symbols
is_allowed_placeholder and scan_for_secrets when making these edits.

**Review by mohavro — COMMENTED** [review]
resolve all comments and merge PR

**Review by mohavro — COMMENTED** [review]
merge PR

File: conftest.py [hunk 7]
```diff
@@ -27,21 +27,25 @@ def _cov_plugin_available() -> bool:
...
-
-    If the plugin is present the argument list is left unchanged.
-    Otherwise, remove occurrences of --cov and --cov-report together
-    with their following parameters.
     Also remove any inline forms starting with --cov= or --cov-report=.
     The original `args` list is updated in-place.
 
     Parameters:
+        early_config: pytest early config object
+        parser: pytest parser object
```
[the remark this answers is not in the document]
Comment (mohavro) [reply conftest.py]:
```suggestion

        parser: pytest parser object
```

File: conftest.py [hunk 8]
```diff
@@ -27,21 +27,25 @@ def _cov_plugin_available() -> bool:
...
-    If the plugin is present the argument list is left unchanged.
-    Otherwise, remove occurrences of --cov and --cov-report together
-    with their following parameters.
-    Also remove any inline forms starting with --cov= or --cov-report=.
-    The original `args` list is updated in-place.
-
+    """
+    Remove pytest - cov related command - line options from the provided,
+    argument list when the pytest - cov plugin is not available.,
+    Also remove any inline forms starting with --cov = or --cov - report = .,
```
[the remark this answers is not in the document]
Comment (mohavro) [reply conftest.py]: @codeant-ai can you review this comment and suggest code to resolve it?
